lib/pac55xx/can.c

```diff
@@ -0,0 +1,515 @@
+  CAN_ISR_SR_CMR_MR_SET(canport, CAN_MR_AFM);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Dual Filter Standard Frame
+Note: Acceptance Code Mask Register values of 1 indicate the filter is to
+ignore the bit. However, standard CAN driver APIs use a positive logic for the
+mask. So this function inverts the mask to make this more portable/intuitive.
+
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] id1 Unsigned int32. CAN ID 1. Only bits 10:0 are used.
+@param[in] id1_mask Unsigned int32. CAN ID 1 mask. Only bits 10:0 are used.
+@param[in] id2 Unsigned int32. CAN ID 2. Only bits 10:0 are used.
+@param[in] id2_mask Unsigned int32. CAN ID 2 mask. Only bits 10:0 are used.
+@param[in] db bool. CAN first data byte value.
+@param[in] db_mask bool. CAN first data byte mask.
+ */
+void can_filter_dual(uint32_t canport, uint32_t id1, uint32_t id1_mask,
+                     uint32_t id2, uint32_t id2_mask,
+                     uint8_t db, uint8_t db_mask) {
+  /* set value */
+  uint32_t word = ((id1 << 21) & CAN_ACR_DUAL_ID1) |
+      ((id2<<5) & CAN_ACR_DUAL_ID2) |
+      ((db<<12) & CAN_ACR_DUAL_DB_UPPER) |
+      (db & CAN_ACR_DUAL_DB_LOWER);
+  CAN_ACR(canport) = __builtin_bswap32(word);
+  /* set mask */
+  word = ((~id1_mask << 21) & CAN_ACR_DUAL_ID1) |
+      ((~id2_mask<<5) & CAN_ACR_DUAL_ID2) |
+      ((~db_mask<<12) & CAN_ACR_DUAL_DB_UPPER) |
+      ((~db_mask) & CAN_ACR_DUAL_DB_LOWER) |
+      CAN_ACR_DUAL_RTR1 |
+      CAN_ACR_DUAL_RTR2;
+  CAN_AMR(canport) = __builtin_bswap32(word);
+  /* 0: dual filter */
+  CAN_ISR_SR_CMR_MR_CLEAR(canport, CAN_MR_AFM);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Filter Single Standard Frame
+Note: Acceptance Code Mask Register values of 1 indicate the filter is to
+ignore the bit. However, standard CAN driver APIs use a positive logic for the
+mask. So this function inverts the mask to make this more portable/intuitive.
+
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] id Unsigned int32. CAN ID. Only bits 10:0 are used.
+@param[in] id_mask Unsigned int32. CAN ID mask. Only bits 10:0 are used.
+@param[in] rtr bool. Remote Request bit value.
+@param[in] rtr_mask bool. Remote Request mask value.
+@param[in] db1 bool. CAN first data byte value.
+@param[in] db1_mask bool. CAN first data byte mask.
+@param[in] db2 bool. CAN second data byte value.
+@param[in] db2_mask bool. CAN second data byte mask.
+ */
+void can_filter_single_std(uint32_t canport, uint32_t id, uint32_t id_mask,
+                                uint8_t db1, uint8_t db1_mask,
+                                uint8_t db2, uint8_t db2_mask) {
+  /* set value */
+  uint32_t word = ((id << 21) & CAN_ACR_SINGLE_STD_ID) |
+      ((db1<<8) & CAN_ACR_SINGLE_STD_DB1) |
+      ((db2<<0) & CAN_ACR_SINGLE_STD_DB2);
+  CAN_ACR(canport) = __builtin_bswap32(word);
```

Comment:
You're making the words from scratch, why do you need to then swap it?

Reply:
Unfortunately, the user guide is messed up.  The macros I made are per the user guide.  However, the register itself is byte-swapped relative to their definition.  We got confirmation from Qorvo in email indicating this swap.  I could go back and change the macros, etc... if you'd prefer, but then it wouldn't "look" like the user guide.

Reply:
And.... they're never going to fix the docs?  This is a pretty absurd situation.  Docs get updated, we're wrong again, no matter what we do. I'm _tempted_ at the moment to say have the definitions be reality, not docs, with comments on the definitions,  then, if the docs get updated to match reality, we have no changes whatsowever, and only people who are looking at the definitions explicitly would ever be confused anyway, if they just use the symbolic names everything "just works" as intended.  I'm not sure what to do here yet. :|

Reply:
Qorvo told us on email "the documentation needs to be corrected".  However, looking at the documentation below, it looks "clean" before doing the byte-swap.  After doing the byte-swap, the documentation could be confusing.  They may just put verbiage in the doc indicating to do a byte-swap for these registers.  Anyway, I wanted you to take a look at this.  My gut is telling me to do the swap in the macros, and go through the headache of mixing up the bits in the function, in order to align it with documentation.  But I think the `__builtin_bswap32()` is the cleanest way to write the code.

If you give me the go-ahead, I'll make the changes to pull out the `__builtin_bswap32()`.

![image](https://user-images.githubusercontent.com/37257349/75448325-fca60200-5938-11ea-84c1-98f8a5047fa7.png)

Reply:
Unfortunately this is a bit of a conundrum, but because the registers are endian swapped, and we would not expect the user to call this function with endian swapped filter inputs, it is actually cleaner, and more efficient to just do the single swap. It also will be more error prone, difficult to read, etc. to do individual swaps because 11-bit ID's will not necessarily cross "natural byte boundaries" (i.e. you may have to shift-and-swap-and-shift to get right value).

I will contact the MFG again and see how they intend to fix the documentation. If their fix is just a note that the registers should be accessed as a BE register, then I think the code as is actually meets that.

Reply:
FYI, just received response from MFG, their thought is the likely update will just be that these registers should be accessed as BE.

Reply:
ok, given that, I guess bswap internally to the library is as rational as things get here then I guess :)

Reply:
cool. I'll resolve this then